File: ABC/abc301-abc350/abc343/f/main.cpp
#include <bits/stdc++.h>
#include <atcoder/segtree>

using namespace std;
using namespace atcoder;

using ll = long long;
const ll INF = 1e18;

// (1st value, 1st count, 2nd value, 2nd count)
struct Node {
    ll first_value;
    ll first_count;
    ll second_value;
    ll second_count;
};

Node op(Node x, Node y) {
    map<ll, ll> value_count;

    // value, countのペアに変換
    value_count[x.first_value] += x.first_count;
    value_count[x.second_value] += x.second_count;
    value_count[y.first_value] += y.first_count;
    value_count[y.second_value] += y.second_count;

    // x, yにおける1st, 2ndのvalueとcountを取得
    vector<pair<ll, ll>> sorted_values(value_count.begin(), value_count.end());
    sort(sorted_values.rbegin(), sorted_values.rend());

    Node result;
    result.first_value = sorted_values[0].first;
    result.first_count = sorted_values[0].second;
    
    if (sorted_values.size() > 1) {
        result.second_value = sorted_values[1].first;
        result.second_count = sorted_values[1].second;
    } else {
        result.second_value = -INF;
        result.second_count = 0;
    }

    return result;
}

Node e() {
    return {-INF, 0, -INF + 1, 0};
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;
    vector<int> a(n);
    
    for (int i = 0; i < n; ++i) {
        cin >> a[i];
    }

    vector<Node> initial(n);
    
    for (int i = 0; i < n; ++i) {
        initial[i] = {a[i], 1, -INF, 0};
    }

    segtree<Node, op, e> seg(initial);

    for (int i = 0; i < q; ++i) {
        int qi;
        cin >> qi;
        
        if (qi == 1) {
            int p, x;
            cin >> p >> x;
            --p;
            seg.set(p, {x, 1, -INF, 0});
        } else {
            int l, r;
            cin >> l >> r;
            --l;
            Node res = seg.prod(l, r);
            cout << res.second_count << '\n';
        }
    }

    return 0;
}

File: ABC/abc301-abc350/abc343/f/main.py
# -*- coding: utf-8 -*-


# PyPy3でもTLE
def main():
    import sys
    from collections import defaultdict

    from atcoder.segtree import SegTree

    input = sys.stdin.readline

    n, q = map(int, input().split())
    a = list(map(int, input().split()))
    inf = 10**18

    # (1st value, 1st count, 2nd value, 2nd count)
    def op(x, y):
        # value, countのペアに変換
        x_1st, x_2nd = (x[0], x[1]), (x[2], x[3])
        y_1st, y_2nd = (y[0], y[1]), (y[2], y[3])

        # 同じvalueのときは、countを合算
        value_count = defaultdict(int)

        for value, count in [x_1st, x_2nd, y_1st, y_2nd]:
            if value in value_count:
                value_count[value] += count
            else:
                value_count[value] = count

        # x, yにおける1st, 2ndのvalueとcountを取得
        value_count = sorted(value_count.items(), reverse=True)

        first_value, first_count = value_count[0]
        second_value, second_count = (
            value_count[1] if len(value_count) > 1 else (-inf, 0)
        )

        return first_value, first_count, second_value, second_count

    e = (-inf, 0, -inf + 1, 0)
    seg = SegTree(op, e, [(ai, 1, -inf, 0) for ai in a])

    for _ in range(q):
        qi, *args = map(int, input().split())

        if qi == 1:
            p, x = args
            p -= 1

            seg.set(p, (x, 1, -inf, 0))
        else:
            l, r = args
            l -= 1

            _, _, _, second_count = seg.prod(l, r)
            print(second_count)


if __name__ == "__main__":
    main()
